src/compute-client/src/as_of_selection.rs

```diff
@@ -435,24 +434,49 @@ impl<'a, T: TimestampManipulation> Context<'a, T> {
 
     /// Apply as-of constraints imposed by the frontiers of downstream storage collections.
     ///
-    /// A collection's as-of _must_ be <= the join of the read frontier and the write frontier of
-    /// the storage collection it exports to, if any.
+    /// A collection's as-of _must_ be < the write frontier of the storage collection it exports to
+    /// (if any) if it is non-empty, and <= the storage collection's read frontier otherwise.
     ///
-    /// We need to consider the read frontier in addition to the write frontier because storage
-    /// collections are commonly initialyzed with a write frontier of `[0]`, even when they start
-    /// producing output at some later time. The read frontier is always initialized to the first
-    /// time the collection will produce valid output for, so it constrains the times that need to
-    /// be produced by dependencies of newly initialized storage collections.
+    /// Rationale:
+    ///
+    /// * A collection's as-of must be <= the write frontier of its dependent storage collection,
+    ///   because we need to pick up computing the contents of storage collections where we left
+    ///   off previously, to avoid skipped times observable in the durable output.
+    /// * Some dataflows feeding into storage collections (specifically: continual tasks) need to
+    ///   be able to observe input changes at times they write to the output. If we selected the
+    ///   as-of to be equal to the write frontier of the output storage collection, we wouldn't be
+    ///   able to produce the correct output at that frontier. Thus the selected as-of must be
+    ///   strictly less than the write frontier.
+    /// * As an exception to the above, if the output storage collection is empty (i.e. its write
+    ///   frontier is <= its read frontier), we need to allow the as-of to be equal to the read
+    ///   frontier. This is correct in the sense that it mirrors the timestamp selection behavior
+    ///   of the sequencer when it created the collection. Chances are that the sequencer chose the
+    ///   initial as-of (and therefore the initial read frontier of the storage collection) as the
+    ///   smallest possible time that can still be read from the collection inputs, so forcing the
+    ///   upper bound any lower than that read frontier would produce a hard constraint violation.
```

Comment:
This part is a bit awkward, but without changing how as-ofs are selected during sequencing, we can't guarantee that `read_frontier - 1` is still valid in the inputs (and in fact most of the time it will not be), so we need the special case.